<a href="https://colab.research.google.com/github/Ahd-Rizlan/Machine-Learning/blob/Notebook-01/Notebook_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Installing Pandas on Notebook-01

In [ ]:
!pip install pandas

Import Required Libraries

In [ ]:
import pandas as pd
#Importing Ploty to plot Diagrams
import plotly.express as px

Loading CSV to DataFrame

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/ML-CW/CourseworkDataset.csv')

Checking the shape - [colums and Rows] of the dataset

In [ ]:
df.shape

(4024, 18)

Checking the Data contents

In [ ]:
#displaying first five rows
df.head()

Patient_ID  Month_of_Birth   Age     Sex   Occupation T_Stage N_Stage  \
0      A0012              12  68.0  Female     Teaching      T1      N1   
1      A0013              12  50.0  Female      Medical      T2      N2   
2      A0014              11  58.0  Female  Engineering      T3      N3   
3      A0015               3  58.0  Female   Technology      T1      N1   
4      A0016               1  47.0  Female   Multimedia      T2      N1   

  6th_Stage             Differentiated  Grade   A_Stage  Tumor_Size  \
0       IIA      Poorly differentiated      3  Regional         4.0   
1      IIIA  Moderately differentiated      2  Regional        35.0   
2      IIIC  Moderately differentiated      2  Regional        63.0   
3       IIA      Poorly differentiated      3  Regional        18.0   
4       IIB      Poorly differentiated      3  Regional        41.0   

  Estrogen_Status Progesterone_Status  Regional_Node_Examined  \
0        Positive            Positive                    24.0   
1        Positive            Positive                    14.0   
2        Positive            Positive                    14.0   
3        Positive            Positive                     2.0   
4        Positive            Positive                     3.0   

   Reginol_Node_Positive  Survival_Months Mortality_Status  
0                      1               60            Alive  
1                      5               62            Alive  
2                      7               75            Alive  
3                      1               84            Alive  
4                      1               50            Alive

List Of Columns in CW dataset

In [ ]:
# Program to print all the column names of the data_frame
list(df.columns)

['Patient_ID',
 'Month_of_Birth',
 'Age',
 'Sex',
 'Occupation',
 'T_Stage',
 'N_Stage',
 '6th_Stage',
 'Differentiated',
 'Grade',
 'A_Stage',
 'Tumor_Size',
 'Estrogen_Status',
 'Progesterone_Status',
 'Regional_Node_Examined',
 'Reginol_Node_Positive',
 'Survival_Months',
 'Mortality_Status']

Taking Summary of cw dataset

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Patient_ID              4024 non-null   object 
 1   Month_of_Birth          4024 non-null   int64  
 2   Age                     4015 non-null   float64
 3   Sex                     4020 non-null   object 
 4   Occupation              43 non-null     object 
 5   T_Stage                 4024 non-null   object 
 6   N_Stage                 4024 non-null   object 
 7   6th_Stage               4024 non-null   object 
 8   Differentiated          4024 non-null   object 
 9   Grade                   4024 non-null   int64  
 10  A_Stage                 4024 non-null   object 
 11  Tumor_Size              4021 non-null   float64
 12  Estrogen_Status         4024 non-null   object 
 13  Progesterone_Status     4024 non-null   object 
 14  Regional_Node_Examined  4023 non-null   

By the summry we have an idea that there are following dataTypes

1.   3 Float
2.   4 Integer
3.  11 Objects



Descriptive statistics of the dataset
like
***For Numeric Data***
1.   count
2.   mean
3.   std
4.   min
5.   max
6.   50
7.   upper percnetiles and etc

For Object Data
1. Count
2. unique
3. top
4. freq

In [ ]:
df.describe()

Month_of_Birth          Age        Grade   Tumor_Size  \
count     4024.000000  4015.000000  4024.000000  4021.000000   
mean         6.481362    54.107098     2.150596    30.419299   
std          3.475442    11.715528     0.638234    21.161080   
min          1.000000   -50.000000     1.000000   -75.000000   
25%          3.000000    47.000000     2.000000    16.000000   
50%          6.000000    54.000000     2.000000    25.000000   
75%         10.000000    61.000000     3.000000    38.000000   
max         12.000000   502.000000     4.000000   140.000000   

       Regional_Node_Examined  Reginol_Node_Positive  Survival_Months  
count             4023.000000            4024.000000      4024.000000  
mean                14.373602               4.158052        71.472167  
std                  8.129293               5.109331        25.361855  
min                  1.000000               1.000000         1.000000  
25%                  9.000000               1.000000        56.000000  
50%                 14.000000               2.000000        73.000000  
75%                 19.000000               5.000000        90.000000  
max                 61.000000              46.000000       760.000000

In [ ]:
df.describe(include='object')

Patient_ID     Sex    Occupation T_Stage N_Stage 6th_Stage  \
count        4024    4020            43    4024    4024      4024   
unique       4024       2            40       4       3         5   
top         A0012  Female  House Person      T2      N1       IIA   
freq            1    4001             2    1786    2732      1305   

                   Differentiated   A_Stage Estrogen_Status  \
count                        4024      4024            4024   
unique                          4         2               2   
top     Moderately differentiated  Regional        Positive   
freq                         2351      3932            3755   

       Progesterone_Status Mortality_Status  
count                 4024             4024  
unique                   2                7  
top               Positive            Alive  
freq                  3326             3399

Number of Null or Empty Values in the Dataset

In [ ]:
df.isna().sum()

Patient_ID                   0
Month_of_Birth               0
Age                          9
Sex                          4
Occupation                3981
T_Stage                      0
N_Stage                      0
6th_Stage                    0
Differentiated               0
Grade                        0
A_Stage                      0
Tumor_Size                   3
Estrogen_Status              0
Progesterone_Status          0
Regional_Node_Examined       1
Reginol_Node_Positive        0
Survival_Months              0
Mortality_Status             0
dtype: int64

by this we have an idea that

*   Occupations have more Empty Value
*   Age, Sex, Tumor_Size, Regional_Node_Examined have some null Values



We have multiple Objects so we have to check the unique values of them

The following funtion will print the Object and Their Unique Value

In [ ]:
def checkUniqueValues(df,ObjectName):
  print("-"*50)
  print(f"Unqiue Values for {ObjectName}")
  print(df[ObjectName].unique())


In [ ]:
for col in df.select_dtypes(include='object').columns:
    checkUniqueValues(df, col)

--------------------------------------------------
Unqiue Values for Patient_ID
['A0012' 'A0013' 'A0014' ... 'A4033' 'A4034' 'A4035']
--------------------------------------------------
Unqiue Values for Sex
['Female' '1' nan]
--------------------------------------------------
Unqiue Values for Occupation
['Teaching' 'Medical' 'Engineering' 'Technology' 'Multimedia' 'Research'
 'House Person' 'Reporting' 'Development' 'Construction' 'Leisure'
 'Scoial work' 'Trading' 'Consultancy' 'Driving' 'Planning' 'Hospitality'
 'Travel' 'Agriculture' 'Manufacturing' 'Photography' 'Finance' 'Business'
 'Marketing' 'Training' 'Cleaning' 'Legal' 'Healthcare' 'Military' 'Sales'
 'Entertainment' 'Sports' 'Arts' 'Design' 'Maintenance' 'Management'
 'Human Resources' 'Administrative' 'Public' 'Politics' nan]
--------------------------------------------------
Unqiue Values for T_Stage
['T1' 'T2' 'T3' 'T4']
--------------------------------------------------
Unqiue Values for N_Stage
['N1' 'N2' 'N3']
-------

Here we can see Patient_ID which is unique for all , so we should remove this

Occupation this has more Null Values more than pradiction its better to remove it

Month_of_Birth is an garbage column

In [ ]:
df = df.drop(columns=(["Patient_ID","Occupation","Month_of_Birth"]),axis=1)

Check Null Values after Removing Columns

In [ ]:
df.isna().sum()

Age                       9
Sex                       4
T_Stage                   0
N_Stage                   0
6th_Stage                 0
Differentiated            0
Grade                     0
A_Stage                   0
Tumor_Size                3
Estrogen_Status           0
Progesterone_Status       0
Regional_Node_Examined    1
Reginol_Node_Positive     0
Survival_Months           0
Mortality_Status          0
dtype: int64

Since we have Plenty of data,
 we can drop Null Values in Gender

Other Values are Filled with Median since

In [ ]:
df = df.dropna(subset=['Sex'])
df.isnull().sum()


Age                       9
Sex                       0
T_Stage                   0
N_Stage                   0
6th_Stage                 0
Differentiated            0
Grade                     0
A_Stage                   0
Tumor_Size                3
Estrogen_Status           0
Progesterone_Status       0
Regional_Node_Examined    1
Reginol_Node_Positive     0
Survival_Months           0
Mortality_Status          0
dtype: int64

Fill the Null Values with Median Value

In [ ]:
df['Age'] = df['Age'].fillna(df['Age'].median())
df['Tumor_Size'] = df['Tumor_Size'].fillna(df['Tumor_Size'].median())
df['Regional_Node_Examined'] = df['Regional_Node_Examined'].fillna(df['Regional_Node_Examined'].median())

df.isnull().sum()/len(df)*100

<ipython-input-504-c849fa358456>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-504-c849fa358456>:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-504-c849fa358456>:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Age                       0.0
Sex                       0.0
T_Stage                   0.0
N_Stage                   0.0
6th_Stage                 0.0
Differentiated            0.0
Grade                     0.0
A_Stage                   0.0
Tumor_Size                0.0
Estrogen_Status           0.0
Progesterone_Status       0.0
Regional_Node_Examined    0.0
Reginol_Node_Positive     0.0
Survival_Months           0.0
Mortality_Status          0.0
dtype: float64

Since the Datas have inconsistency in Vales, we have to order them

In [ ]:
def changeValuesInGoodOrder(Value):
  if Value.lower() == "alive":
    return "Alive"
  else:
    return "Dead"

In [ ]:
df['Mortality_Status'] = df['Mortality_Status'].apply(changeValuesInGoodOrder)
print(df['Mortality_Status'].unique())

['Alive' 'Dead']


<ipython-input-506-aae89218afc3>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Assuming 1 as Males and naming them correctly

In [ ]:
def orderTheSexData(Value):
  if str(Value).lower() == 'female':
    return "Female"
  elif Value == '1' or Value == 1:
    return 'Male'
  else :
    return None

In [ ]:
df['Sex'] = df['Sex'].apply(orderTheSexData)
print(df['Sex'].unique())

['Female' 'Male']


<ipython-input-508-b274c04b5e4a>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
df['Sex'].value_counts()

Sex
Female    4001
Male        19
Name: count, dtype: int64

Here you can see there are just 19 data , this might cost a noise so better to get rid of this sex column


In [ ]:
df = df.drop(columns=(["Sex"]),axis=1)

There are No Null Values, so Next step is to find Outliers


In [ ]:
df.describe().transpose()

count       mean        std   min   25%   50%   75%  \
Age                     4020.0  54.113682  11.706125 -50.0  47.0  54.0  61.0   
Grade                   4020.0   2.151493   0.637772   1.0   2.0   2.0   3.0   
Tumor_Size              4020.0  30.396766  21.144770 -75.0  16.0  25.0  38.0   
Regional_Node_Examined  4020.0  14.375124   8.130013   1.0   9.0  14.0  19.0   
Reginol_Node_Positive   4020.0   4.157463   5.110944   1.0   1.0   2.0   5.0   
Survival_Months         4020.0  71.463184  25.371926   1.0  56.0  73.0  90.0   

                          max  
Age                     502.0  
Grade                     4.0  
Tumor_Size              140.0  
Regional_Node_Examined   61.0  
Reginol_Node_Positive    46.0  
Survival_Months         760.0

From the above we can undestand that
1. Age
2. Tumor_Size
3. RNE
4. RNP
5. Survival Month

this can be seen because min can have (-) value , very high Max Value

*Below funtion will Print Histogram, Scatter, Box for Various Values*

In [ ]:
def printOutlierVisalization(Value):
  print (f"Ourlier Visulaization for {Value}")
  ShowHistogram = px.histogram(df, x=Value).show()
  ShowScatter = px.scatter(df, x=Value).show()
  ShowBox = px.box(df, x=Value).show()

Oulier Visualization For Age

In [ ]:
printOutlierVisalization("Age")

Ourlier Visulaization for Age


The Above Diagrams shows that there are Extreme values , which we have to getrid of

Todo that will use  find outliers we can use a method called IQR statistical

it will use 1st Quater and 3rd Quator and Substracts them and get the IQRValue then

find the values which are out of (1st Quater-(1.5IQRValue) or 3rd Quater+(1.5IQRValue)

In [ ]:
def find_outliers_IQR(df):
  q1= df.quantile(0.25)
  q3 = df.quantile(0.75)
  IQR = q3 - q1
  outliers = df[((df<(q1-1.5*IQR))|(df>(q3+1.5*IQR)))]
  return outliers.index



---

Oulier Visualization For **Tumor_Size**

In [ ]:
print(find_outliers_IQR(df['Age']))

Index([142, 212, 522, 842], dtype='int64')


*Here you can see there are 4 extreme values for Age*

*we can remove those by Index*

In [ ]:
df.drop(index=find_outliers_IQR(df['Age']), inplace=True)



---

Oulier Visualization For **Tumor_Size**

In [ ]:
printOutlierVisalization("Tumor_Size")
print(find_outliers_IQR(df['Tumor_Size']))

Ourlier Visulaization for Tumor_Size


Index([   8,   52,   53,   63,   70,   86,   87,  116,  147,  155,
       ...
       3893, 3902, 3909, 3910, 3925, 3930, 3965, 3991, 4009, 4012],
      dtype='int64', length=221)


*Here you can see there are 221 extreme values for Tumor_Size*

In [ ]:
df.drop(index=find_outliers_IQR(df['Tumor_Size']), inplace=True)



---

Oulier Visualization For **Regional_Node_Examined**

In [ ]:
printOutlierVisalization("Regional_Node_Examined")
print(find_outliers_IQR(df['Regional_Node_Examined']))

Ourlier Visulaization for Regional_Node_Examined


Index([  57,  227,  290,  303,  394,  418,  490,  530,  550,  633,  662,  701,
        790,  838,  941,  989, 1027, 1128, 1267, 1459, 1520, 1560, 1580, 1591,
       1624, 1872, 2056, 2128, 2134, 2142, 2159, 2182, 2213, 2243, 2248, 2293,
       2323, 2376, 2460, 2462, 2603, 2607, 2725, 2753, 2887, 2934, 2975, 3043,
       3064, 3097, 3132, 3149, 3158, 3218, 3288, 3326, 3438, 3454, 3493, 3699,
       3708, 3771, 3923, 3950, 3957, 3990],
      dtype='int64')


In [ ]:
df.drop(index=find_outliers_IQR(df['Regional_Node_Examined']), inplace=True)



---

Oulier Visualization For **Reginol_Node_Positive**

In [ ]:
printOutlierVisalization("Reginol_Node_Positive")
print(find_outliers_IQR(df['Reginol_Node_Positive']))

Ourlier Visulaization for Reginol_Node_Positive


Index([   9,   14,   24,   25,   26,   34,   40,   48,   66,   71,
       ...
       3949, 3960, 3967, 3974, 3982, 3985, 3992, 3995, 3996, 4004],
      dtype='int64', length=449)


In [ ]:
df.drop(index=find_outliers_IQR(df['Reginol_Node_Positive']), inplace=True)



---

Oulier Visualization For **Survival Months**

In [ ]:
printOutlierVisalization("Survival_Months")
print(find_outliers_IQR(df['Survival_Months']))

Ourlier Visulaization for Survival_Months


Index([ 414,  520,  631,  679,  843,  928, 1042, 1143, 1156, 1444, 1702, 1709,
       1751, 1893, 2410, 3474, 3711, 3803],
      dtype='int64')


In [ ]:
df.drop(index=find_outliers_IQR(df['Survival_Months']), inplace=True)



---


---




To check wather we have any more Outliers we can describe our dataset again

In [ ]:
df.describe().transpose()


count       mean        std   min   25%   50%   75%  \
Age                     3262.0  53.958001   8.910631  30.0  47.0  54.0  61.0   
Grade                   3262.0   2.116493   0.629377   1.0   2.0   2.0   3.0   
Tumor_Size              3262.0  25.810546  14.449703   1.0  15.0  22.0  32.0   
Regional_Node_Examined  3262.0  13.036787   7.116966   1.0   8.0  13.0  18.0   
Reginol_Node_Positive   3262.0   2.408338   1.809893   1.0   1.0   2.0   3.0   
Survival_Months         3262.0  72.889025  21.503498   6.0  58.0  74.0  91.0   

                          max  
Age                      69.0  
Grade                     4.0  
Tumor_Size               70.0  
Regional_Node_Examined   34.0  
Reginol_Node_Positive     8.0  
Survival_Months         107.0



---


Exporting Dataset for For Mortality_Status  Prediction


---



In [ ]:
df.to_csv(r'/content/drive/MyDrive/ML-CW/Mortality_Status.csv',index=False)



---


Exporting Dataset for For Survival_Months  Prediction


---



For this Im dropping the All the Alive Personals so i can assume that all are dead

so itll be easy to predict

In [ ]:
df['Mortality_Status'].value_counts()


Mortality_Status
Alive    2893
Dead      369
Name: count, dtype: int64

In [ ]:
# Remove alive people from the dataset
df = df[df['Mortality_Status'] != 'Alive'].reset_index(drop=True)

# Drop the Mortality_Status column
df.drop(columns=['Mortality_Status'], inplace=True)

# Check the final structure
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 369 entries, 0 to 368
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     369 non-null    float64
 1   T_Stage                 369 non-null    object 
 2   N_Stage                 369 non-null    object 
 3   6th_Stage               369 non-null    object 
 4   Differentiated          369 non-null    object 
 5   Grade                   369 non-null    int64  
 6   A_Stage                 369 non-null    object 
 7   Tumor_Size              369 non-null    float64
 8   Estrogen_Status         369 non-null    object 
 9   Progesterone_Status     369 non-null    object 
 10  Regional_Node_Examined  369 non-null    float64
 11  Reginol_Node_Positive   369 non-null    int64  
 12  Survival_Months         369 non-null    int64  
dtypes: float64(3), int64(3), object(7)
memory usage: 37.6+ KB


In [ ]:
df.to_csv(r'/content/drive/MyDrive/ML-CW/Survival_Months.csv',index=False)